lib/galaxy/webapps/galaxy/api/users.py

```diff
@@ -240,3 +250,353 @@ def anon_user_api_value( self, trans ):
+
+    @expose_api
+    def toolbox_filters(self, trans, id, payload={}, **kwd):
+        """
+        API call for fetching toolbox filters data. Toolbox filters are specified in galaxy.ini.
+        The user can activate them and the choice is stored in user_preferences.
+        """
+        user = self._get_user(trans, id)
+        filter_types = odict([('toolbox_tool_filters', {'title': 'Tools', 'config': trans.app.config.user_tool_filters}),
+                              ('toolbox_section_filters', {'title': 'Sections', 'config': trans.app.config.user_section_filters}),
+                              ('toolbox_label_filters', {'title': 'Labels', 'config': trans.app.config.user_label_filters})])
+        if trans.get_request_method() == 'PUT':
+            for filter_type in filter_types:
+                new_filters = []
+                for prefixed_name in payload:
+                    prefix = filter_type + '|'
+                    if prefixed_name.startswith(filter_type):
+                        new_filters.append(prefixed_name[len(prefix):])
+                user.preferences[filter_type] = ','.join(new_filters)
+            trans.sa_session.add(user)
+            trans.sa_session.flush()
+            return {'message': 'Toolbox filters have been saved.'}
+        else:
+            saved_values = {}
+            for name, value in user.preferences.items():
+                if name in filter_types:
+                    saved_values[name] = listify(value, do_strip=True)
+            inputs = []
+            factory = FilterFactory(trans.app.toolbox)
+            for filter_type in filter_types:
+                self._add_filter_inputs(factory, filter_types, inputs, filter_type, saved_values)
+            return {'message': 'Toolbox filters unchanged.', 'inputs': inputs}
+
+    def _add_filter_inputs(self, factory, filter_types, inputs, filter_type, saved_values):
+        filter_inputs = list()
+        filter_values = saved_values.get(filter_type, [])
+        filter_config = filter_types[filter_type]['config']
+        filter_title = filter_types[filter_type]['title']
+        for filter_name in filter_config:
+            function = factory.build_filter_function(filter_name)
+            filter_inputs.append({
+                'type': 'boolean',
+                'name': filter_name,
+                'label': filter_name,
+                'help': docstring_trim(function.__doc__) or 'No description available.',
+                'value': 'true' if filter_name in filter_values else 'false',
+                'ignore': 'false'
+            })
+        if filter_inputs:
+            inputs.append({'type': 'section', 'title': filter_title, 'name': filter_type, 'expanded': True, 'inputs': filter_inputs})
+
+    @expose_api
+    def api_key(self, trans, id, payload={}, **kwd):
+        """
+        Get/Create API key.
+        """
+        user = self._get_user(trans, id)
+        if trans.get_request_method() == 'POST':
```

Comment:
(discussed this among other stuff directly with guerler, but commenting to document it here) I think it's more standard (and cleaner) to divert HTTP methods at the routing layer instead of here.